Issue/nav redesign

nbradbury opened the pull request:
This PR updates the app to a new, cleaner design. 

![light](https://user-images.githubusercontent.com/3903757/98254632-87201600-1f4a-11eb-84c4-72e795014d3b.gif)

![dark](https://user-images.githubusercontent.com/3903757/98254641-8ab39d00-1f4a-11eb-83c3-a498e186ac72.gif)

[screenshots.zip](https://github.com/woocommerce/woocommerce-android/files/5497856/screenshots.zip)

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

nbradbury committed:
Changed toolbar and tab colors

nbradbury committed:
Removed toolbar elevation

nbradbury committed:
Changed status bar color

nbradbury committed:
Changed icon colors

nbradbury committed:
Changed overflow menu color

nbradbury committed:
First pass at collapsing toolbar

nbradbury committed:
Set the collpasing toolbar title when activity title is changed

nbradbury committed:
always collapse the AppBar when entering a child fragment

nbradbury committed:
Collapse toolbar when entering a child fragment

nbradbury committed:
Remember if AppBar is expanded

nbradbury committed:
Removed unused import

nbradbury committed:
Toolbars now show correctly in secondary screens

nbradbury committed:
Make sure status bar appears correctly in both light and dark mode

nbradbury committed:
Collapse toolbar when search view appears, re-expand when it disappears

nbradbury committed:
Animate the toolbar change

nbradbury committed:
Clear order list menu before populating it

nbradbury committed:
Merge branch 'develop' of https://github.com/woocommerce/woocommerce-android into issue/nav-redesign

nbradbury committed:
Allow expanded toolbar to wrap two lines

nbradbury committed:
Fixed isScrolledToTop()

nbradbury committed:
Merge branch 'develop' of https://github.com/woocommerce/woocommerce-android into issue/nav-redesign

nbradbury committed:
turn off toolbar animation when leavimg top-level fragment

nbradbury committed:
turn off toolbar animation when leavimg top-level fragment (fixed)

nbradbury committed:
Set expanded toolbar text size to match design

nbradbury committed:
Removed unused style attributes and updated release notes

Garance91540 commented:
Hi Nick! looking 🔥🔥🔥

Could we make the following updates?

**Home Screen - Default** 

Few updates:
- The top nav seems a bit too big, the icons "bar" taking too much space I think (total nav bar height without the tabs should be 108px)
- Vertically center the icons between the text and the status bar
- Make sure the icon right padding is 16px
- Add drop shadow to tab bar

<img width="864" alt="Screen Shot 2020-11-06 at 10 00 17 AM" src="https://user-images.githubusercontent.com/28450084/98346985-1430a100-2017-11eb-81d1-6d332987b3fc.png">

Is it possible to send me a screenshot of this screen for a very store with a very long name?

**Home Screen - Scrolled view** 

Few updates:
- Update font to Headline 6
- Adjust top nav bar (without tabs) height to 56px
- Vertically align the icons in the nav bar

<img width="829" alt="Screen Shot 2020-11-06 at 10 12 41 AM" src="https://user-images.githubusercontent.com/28450084/98348337-e3ea0200-2018-11eb-9a5c-f615d1ab618f.png">

**Product List** 

- Same feedback as for the Home screen top nav
- Check that the filter boxes have 12px top/bottom padding only

<img width="606" alt="Screen Shot 2020-11-06 at 10 22 30 AM" src="https://user-images.githubusercontent.com/28450084/98349314-1fd19700-201a-11eb-8ed6-f88e436e6cae.png">

**Product List - Scrolled** 

- "Filter" and "Sort" buttons collapse on scroll
- Remove drop shadow on the nav bar

<img width="867" alt="Screen Shot 2020-11-06 at 10 53 45 AM" src="https://user-images.githubusercontent.com/28450084/98352546-75a83e00-201e-11eb-87a8-5efac0ac85e1.png">


**Product Details** 

- Update back arrow color to black. I know all buttons should be pink, but as the back arrow + header component can be seen as part of the same entity, we should make them the same color so not to make them stand apart (just a bit more consistent to the eye)
- Adjust top nav bar height to 56px
- Add divider
- Vertically align header and icon

<img width="741" alt="Screen Shot 2020-11-06 at 10 44 40 AM" src="https://user-images.githubusercontent.com/28450084/98351517-1990ea00-201d-11eb-9f45-734a381bb06c.png">

**Settings** 

<img width="803" alt="Screen Shot 2020-11-06 at 10 56 08 AM" src="https://user-images.githubusercontent.com/28450084/98352735-b4d68f00-201e-11eb-971f-e8f992ea350a.png">
 
**tablet**

Could you please send me a screenshot of the tablet view too? 

Thank you! Let me know if you have questions/feedback!

Garance91540 reviewed (commented):
^

nbradbury committed:
Changed expanded toolbar height to 108dp

nbradbury committed:
Set expanded toolbar maxLines to 2

nbradbury committed:
Added end margin to expanded toolbar

nbradbury committed:
Inherit toolbar style from Toolbar.Surface

nbradbury committed:
Added 4dp end padding to toolbar so toolbar icon would be 16dp from edge (icon itself has 12dp)

nbradbury committed:
Allow toolbar to wrap to two lines

nbradbury committed:
Changed end margin of overflow button to 16dp

nbradbury committed:
Merge branch 'develop' of https://github.com/woocommerce/woocommerce-android into issue/nav-redesign

nbradbury commented:
@Garance91540 Thanks for the review! I'll work on those changes and hopefully get this all rolled up early this week. To answer your question, this is what the toolbar looks like with a long store name. I followed the [Material Guidelines](https://material.io/components/app-bars-top#anatomy) regarding app bars.

![two-lines](https://user-images.githubusercontent.com/3903757/98560720-104e8a00-2276-11eb-8fac-65ae1066f73f.png)

Also, the shadow under the tabs appears once you scroll, which is consistent with how the order list works.

![scrolled](https://user-images.githubusercontent.com/3903757/98560839-2fe5b280-2276-11eb-85f2-d8ab9e7fde9c.png)

And here's a tablet view:

![tablet](https://user-images.githubusercontent.com/3903757/98561380-d631b800-2276-11eb-8f98-fcf300c4c30b.png)

nbradbury committed:
Changed toolbar height to 56dp

nbradbury committed:
Added extension to determine if a recycler is scrolled to the top

nbradbury committed:
Rely on computeVerticalScrollOffset to determine whether recycler is scrolled

nbradbury committed:
Make filter card hide on scroll

nbradbury committed:
Changed animation

nbradbury committed:
Added top divider to settings

nbradbury committed:
Removed filter/sort animation

nbradbury committed:
Set the toolbar elevation based on whether a top level fragment is showing

nbradbury commented:
@Garance91540 I believe I've addressed most of your suggestions apart from the following:

- Product list > "Filter" and "Sort" buttons collapse on scroll - I was not able to find a suitable way to do this because those buttons aren't part of the scrolling view. If it's okay with you, I'd like to open a separate issue for that and come back to it.

- Product list > Remove drop shadow on the nav bar - The shadow is only there when the toolbar is collapsed (it's removed when expanded), similar to "My Site" in WPAndroid.

A zip containing full-sized screenshots is below.

[Archive.zip](https://github.com/woocommerce/woocommerce-android/files/5513218/Archive.zip)

Garance91540 commented:
Hi Nick! Thank you for your updates :D

Is it possible to send me the following screenshots too?
- Home screen, scrolled
- Product list, scrolled
- Order list
- Order list, scrolled

> To answer your question, this is what the toolbar looks like with a long store name.

Thanks for the screenshot! I think it will make the interaction between tabs smoother if we keep the top nav consistently the same height, so instead of wrapping the text in several lines, can we truncate after a certain character limit and add "..." so it stays on one line? This is only temporary since we'll update this to "My Store" with the Home screen project 

- Home Screen

<img width="902" alt="Screen Shot 2020-11-10 at 6 33 59 PM" src="https://user-images.githubusercontent.com/28450084/98709995-665d2380-2383-11eb-86f5-3365c50b4872.png">

**Tab bar**

Can we update the tab bar height to 48px like on the designs (i'm using the Google Material tab component). Also, can we update the tab font to "Button"?


> Also, the shadow under the tabs appears once you scroll, which is consistent with how the order list works.

We need a visual divider to avoid [tabs floating on the screen](https://d.pr/i/ZEC8OM). On the designs currently, both scrolled and unscrolled views for tabs layout display drop-shadows ([#](https://d.pr/i/l5Bigi) [#](https://d.pr/i/nw4wt2)). I'm using the material components for tabs. Is it possible to update this? 

**Top nav**

In the screenshot, I highlighted how the padding between the designs and the screenshots differ currently. The top nav should be 108px height, and the top nav icon should have 16px top and bottom padding. The icon should also bee 24px x 24px

**Product list**

> Product list > "Filter" and "Sort" buttons collapse on scroll - I was not able to find a suitable way to do this because those buttons aren't part of the scrolling view. If it's okay with you, I'd like to open a separate issue for that and come back to it.

Sounds good!

> Product list > Remove drop shadow on the nav bar - The shadow is only there when the toolbar is collapsed (it's removed when expanded), similar to "My Site" in WPAndroid.

[In the new designs](https://d.pr/i/tvBc0t), there's no drop shadow on scroll for the top level menu items (except when there's a tab bar). Is it possible to update accordingly? 

**Product Details**

<img width="884" alt="Screen Shot 2020-11-10 at 6 45 18 PM" src="https://user-images.githubusercontent.com/28450084/98711151-ecc63500-2384-11eb-97e8-5c54d2a7ee81.png">

I'm still seeing the following, not sure if you had the chance to address them?
- Header should be Headline 6
- Back arrow color #000000

Thanks for adding the divider! On the screenshot I'm seeing a drop shadow, but we should just see a straight flat line.

**Settings**

Is it possible to update the back arrow color to #000000 ?
Is the header font size "Headline 6"? it seems smaller on the screenshot but it could just be the device!

Thanks Nick! I know these notes may seem like small details, but they will be the visual framework of the app moving forward that's why making them pixel-perfect is quite important 👯‍♀️ 

let me know if you have any questions! I'm afk tomorrow but will try to answer anyway so not to block you

nbradbury committed:
Changed appbar elevation

nbradbury committed:
Merge branch 'develop' of https://github.com/woocommerce/woocommerce-android into issue/nav-redesign

nbradbury committed:
Only allow single-line titles

nbradbury commented:
> Is it possible to send me the following screenshots too?
> 
> * Home screen, scrolled
> * Product list, scrolled
> * Order list
> * Order list, scrolled

@Garance91540 Here you go! 

[Archive.zip](https://github.com/woocommerce/woocommerce-android/files/5519627/Archive.zip)

nbradbury commented:
> let me know if you have any questions! I'm afk tomorrow but will try to answer anyway so not to block you

No worries, enjoy your AFK - this PR can't be merged until next week so you won't be blocking me.

nbradbury committed:
Don't bold collapsed toolbar title

nbradbury committed:
Changed back arrow color

nbradbury committed:
Removed the app bar elevation

nbradbury committed:
Set toolbar elevation based on whether active fragment has tabs

nbradbury committed:
Simplified showing elevation for tabbed fragments

nbradbury committed:
Simplified showing elevation for tabbed fragments

nbradbury committed:
Changed the back button color in settings

nbradbury commented:
> Can we update the tab bar height to 48px like on the designs (i'm using the Google Material tab component). Also, can we update the tab font to "Button"?

I've updated the tab height. The tab font is already Button, though.

> In the screenshot, I highlighted how the padding between the designs and the screenshots differ currently. The top nav should be 108px height, and the top nav icon should have 16px top and bottom padding. The icon should also bee 24px x 24px

The icon padding is the only thing I haven't been able to figure out. That's a toolbar control and it's in charge of placing the icons. I asked other Android devs about this, and hopefully they'll know how to do this.

> We need a visual divider to avoid tabs floating on the screen. On the designs currently, both scrolled and unscrolled views for tabs layout display drop-shadows (# #). I'm using the material components for tabs. Is it possible to update this?

Okay, this is updated.

> In the new designs, there's no drop shadow on scroll for the top level menu items (except when there's a tab bar). Is it possible to update accordingly?

Done!

> Header should be Headline 6

Fixed (I was using Headline 6, but incorrectly bolded it as well). I've corrected it.

> Back arrow color #000000

Fixed.

New screenshots below.

[Archive.zip](https://github.com/woocommerce/woocommerce-android/files/5526769/Archive.zip)

Garance91540 commented:
Thanks Nick!

> The icon padding is the only thing I haven't been able to figure out. That's a toolbar control and it's in charge of placing the icons. I asked other Android devs about this, and hopefully they'll know how to do this.

thanks!

Everything looks good on my end, the only thing missing on top of the top nav icon padding it the collapsing of the "Filer" and "Sort" buttons on scroll that we can address in a separate issue (I see WP Android has a similar behavior, maybe they have some insights?)

nbradbury commented:
@Garance91540 I got some help from @AmandaRiu on the position of the settings icon, and it turns out it's already perfectly centered vertically in the toolbar. We're wondering if maybe you're including the Android system bar in your measurements?

![Screenshot_1605554182](https://user-images.githubusercontent.com/3903757/99298008-0ea53900-2817-11eb-8d7a-23c87d3c5b8c.png)

nbradbury committed:
Merge branch 'develop' of https://github.com/woocommerce/woocommerce-android into issue/nav-redesign

nbradbury committed:
Updated release notes

Garance91540 commented:
Hi @nbradbury ! I think what's happening is that we are not using the same status bar https://d.pr/i/8q9D1n

What's important is that the icon is centered no matter what the status bar is, so it seems like it's good to go!

Garance91540 commented:
I somehow requested a new review from myself but please disregard it!

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt:
```diff
@@ -434,9 +449,28 @@ class MainActivity : AppUpgradeActivity(),
+        if (isAtRoot && isToolbarExpanded) {
+            app_bar_layout.setExpanded(true, true)
+        } else if (!isAtRoot) {
+            app_bar_layout.setExpanded(false, false)
```
np: recommend using `expandToolbar(..)` method below for consistency so if changes are needed later it can all happen in a single place.

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt:
```diff
@@ -894,4 +932,19 @@ class MainActivity : AppUpgradeActivity(),
+     * enabling us to set the elevation when added so there's a shadow under it. Note that we delay adding the
+     * elevation because setting it immediately after the tabLayout is added has no effect.
+     */
+    override fun onChildViewAdded(parent: View?, child: View?) {
+        parent?.postDelayed({
+            app_bar_layout.elevation = resources.getDimensionPixelSize(R.dimen.appbar_elevation).toFloat()
+        }, 100L)
+    }
```
Slick 👍

AmandaRiu commented on WooCommerce/src/main/res/values/dimens_base.xml:
```diff
@@ -105,7 +105,9 @@ so that's the baseline as defined in `major_100`.
     <dimen name="margin_app_title_aligned">72dp</dimen>
     <dimen name="min_tap_target">48dp</dimen>
-    <dimen name="appbar_elevation">4dp</dimen>
+    <dimen name="appbar_elevation">2dp</dimen>
```
Just curious why the change to the toolbar height?

AmandaRiu commented on WooCommerce/src/main/res/values/styles_base.xml:
```diff
@@ -11,16 +11,44 @@ theme across the entire app. Overridden versions should be added to the styles.x
+        <item name="android:textStyle">bold</item>
+        <item name="android:layout_height">@dimen/expanded_toolbar_height</item>
+    </style>
+    <style name="Woo.TextAppearance.CollapsingToolbar.Collapsed" parent="TextAppearance.MaterialComponents.Headline6">
```
np: The current naming convention for custom `TextAppearance` styles follows the way Android does it with `TextAppearance.Woo`.

AmandaRiu reviewed (changes requested):
@nbradbury  Code changes look really good. I just left some nitpicks for your consideration. I found the following UI items during testing:

1. **Detail views**: Should the toolbar in detail views also use collapsing toolbars? I noticed the toolbar will expand in order and review details, but not in product details. 

Order | Review | Product
-- | -- | --
![Screenshot_1605656959](https://user-images.githubusercontent.com/5810477/99465089-936f8000-2907-11eb-9253-cb9486d8415b.png)|![Screenshot_1605657003](https://user-images.githubusercontent.com/5810477/99465099-966a7080-2907-11eb-8c19-c70ab223d8ab.png)|![Screenshot_1605657850](https://user-images.githubusercontent.com/5810477/99465149-aeda8b00-2907-11eb-978c-a50b7a0474fd.png)

2. **Webview**: The color for light mode is still purple

<img src="https://user-images.githubusercontent.com/5810477/99465211-df222980-2907-11eb-9218-fbdc0fc56b10.png" width="300"/>

3. **White back arrow**: The "back" arrow is white in light mode for the Application Log screen and Help & Support screen: 

App Log | Help 
-- | --
![Screenshot_1605656779](https://user-images.githubusercontent.com/5810477/99465404-5952ae00-2908-11eb-9dbb-6695ea06c555.png)|![Screenshot_1605656790](https://user-images.githubusercontent.com/5810477/99465410-5c4d9e80-2908-11eb-9a04-f8da109208ce.png)

4. **Zendesk** screens have wonky styles and need new white appbar styling. Styling these screens is a PITA - [look at this PR for some pointer](https://github.com/woocommerce/woocommerce-android/pull/2170)s. 

Tickets | Contact us 
-- | --
![Screenshot_1605656767](https://user-images.githubusercontent.com/5810477/99465602-c5cdad00-2908-11eb-8e24-6efcdf428c02.png)|![Screenshot_1605656761](https://user-images.githubusercontent.com/5810477/99465607-c8c89d80-2908-11eb-9ceb-992ef499fa49.png)

nbradbury committed:
Merge branch 'develop' into issue/nav-redesign

nbradbury committed:
Fix back button in log viewer

nbradbury committed:
Fix back button in help

nbradbury committed:
Fixed ChromeCustomTab toolbar color

nbradbury committed:
Use expandToolbar

nbradbury committed:
First pass at disabling toolbar expansion

nbradbury committed:
2nd pass at disabling toolbar expansion

nbradbury committed:
Simplified disabling toolbar expansion

nbradbury committed:
Fixed style issues

nbradbury committed:
Restored appbar_elevation to previous value

nbradbury committed:
Fixed TextAppearance style names

nbradbury commented:
@AmandaRiu Thanks for the thorough review and for catching so many things! I've fixed everything except these two, which for now have me stumped:

1. Fixing the Zendesk screens - the Zendesk layouts set their own theme and I don't see how to override the toolbar and toolbar text colors without changing `colorPrimary`, etc., which looks even weirder. I've looked at the Zendesk theme PR you linked to above and read their theming article, but I don't see how to change the toolbar. Any tips?

<img width="1017" alt="zen" src="https://user-images.githubusercontent.com/3903757/99585694-a8eeb380-29b4-11eb-9d08-cb42195f81fe.png">

2. The toolbar should not be expandable in child fragments. I've fixed this, but when you PTR there's now an empty space. I'm still working on this.

![pull](https://user-images.githubusercontent.com/3903757/99585854-e5221400-29b4-11eb-95df-2b3bf1710060.png)

nbradbury committed:
Added comment

AmandaRiu commented:
@nbradbury 

> Fixing the Zendesk screens - the Zendesk layouts set their own theme and I don't see how to override the toolbar and toolbar text colors without changing colorPrimary, etc., which looks even weirder. I've looked at the Zendesk theme PR you linked to above and read their theming article, but I don't see how to change the toolbar. Any tips?

I tried a couple things but didn't have much success. The way Zendesk used styles and themes is terrible - they hard-coded so much! You could try overriding `ThemeOverlay.MaterialComponents.Dark.ActionBar` by creating a style named exactly the same (I did that with login styles), or you could duplicate that activity and style it directly. If the layout file is named exactly the same it will take precedence over the one in the library. The problem there of course is maintaining changes like that. I was doing everything I could to avoid it.

nbradbury committed:
Renamed back icon

nbradbury committed:
Set toolbar icon directly

nbradbury committed:
Override Zendesk layouts so we can apply our own toolbar style

nbradbury committed:
Renamed cross icon and changed color to on surface

nbradbury committed:
Fixed fullscreen image viewer

nbradbury commented:
> ...you could duplicate that activity and style it directly. If the layout file is named exactly the same it will take precedence over the one in the library. The problem there of course is maintaining changes like that. I was doing everything I could to avoid it.

@AmandaRiu In the end, this was the only thing that worked for me. I hate it, but nothing else worked and we couldn't leave the screens the way they were.

I made that change and also fixed a problem in the full-screen image viewer (it was incorrectly showing the collapsible toolbar). Ready for another round!

AmandaRiu reviewed (approved):
Nice work @nbradbury!! 🎉 I know from personal experience how frustrating some of the style stuff can be. The only thing outstanding is the way the navbar will expand on _some_ child fragments -  though it will only show an empty space. Per our slack conversation you'll be addressing that in a separate PR. I tested through as many screens as I could think of and created this list of views that still have the expanding empty toolbar issue:

Expanding toolbar on child fragments (doesn’t happen on all screens with PTR functionality)
	- Product Detail -> short description 
	- Product detail -> description
	- Product detail -> photos grid screen
	- Order detail
	- Order detail -> issue refund
	- Order detail -> add tracking
	- Review detail


Everything else looks crisp, clean and purrrty. :shipit:
